charmcraft/charm_builder.py

```diff
@@ -0,0 +1,330 @@
+
+    def handle_dependencies(self):
+        """Handle from-directory and virtualenv dependencies."""
+        logger.debug("Installing dependencies")
+
+        # virtualenv with other dependencies (if any)
+        if self.requirement_paths:
+            retcode = _process_run(["pip3", "--version"])
+            if retcode:
+                raise CommandError("problems using pip")
+
+            venvpath = self.buildpath / VENV_DIRNAME
+            cmd = [
+                "pip3",
+                "install",  # base command
+                "--target={}".format(
+                    venvpath
+                ),  # put all the resulting files in that specific dir
+            ]
+            if _pip_needs_system():
+                logger.debug("adding --system to work around pip3 defaulting to --user")
+                cmd.append("--system")
+            for reqspath in self.requirement_paths:
+                cmd.append(
+                    "--requirement={}".format(reqspath)
+                )  # the dependencies file(s)
+            retcode = _process_run(cmd)
+            if retcode:
+                raise CommandError("problems installing dependencies")
+
+
+def _pip_needs_system():
+    """Determine whether pip3 defaults to --user, needing --system to turn it off."""
+    cmd = [
+        "python3",
+        "-c",
+        (
+            "from pip.commands.install import InstallCommand; "
+            'assert InstallCommand().cmd_opts.get_option("--system") is not None'
+        ),
+    ]
+    proc = subprocess.run(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
+    return proc.returncode == 0
+
+
+def _process_run(cmd, **kwargs) -> int:
+    logger.debug("Running external command %s", cmd)
+    try:
+        ret = subprocess.run(cmd, **kwargs)
+    except Exception as err:
+        logger.error("Executing %s crashed with %r", cmd, err)
+        return 1
+
+    return ret.returncode
+
+
+def _parse_arguments() -> argparse.Namespace:
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--entrypoint",
+        metavar="filename",
+        default="src/charm.py",
+        help="The charm entry point. Default is 'src/charm.py'.",
+    )
+    parser.add_argument(
+        "--charmdir",
+        metavar="dirname",
+        default=".",
+        help="The charm source directory. Default is current.",
+    )
+    parser.add_argument(
+        "--builddir",
+        metavar="dirname",
+        required=True,
+        help="The build destination directory",
+    )
+    parser.add_argument(
+        "-r",
+        "--requirement",
+        metavar="reqfile",
+        action="append",
+        default=None,
+        help="Comma-separated list of requirements files.",
+    )
+
+    return parser.parse_args()
+
+
+def main():
+    """Run the command-line interface."""
+    options = _parse_arguments()
+
+    logging.basicConfig(level=logging.DEBUG, format="%(message)s")
```

Comment:
Please add a timestamp and a "source" to each log line.

Reply:
If we do that, this information will be duplicated in the charmcraft log, because the tool output will be captured in the parent process logger and the timestamp will be added again. The way it is, the final log appears as:
```
2021-07-20 09:39:14,290  charmcraft.commands.build                DEBUG    Building for 'bases[0]' as host matches 'build-on[0]'.                                                      
2021-07-20 09:39:14,291  charmcraft.commands.build                DEBUG    Building charm in '/home/claudio/charmcraft/charmcraft/testdir/build'                                       
2021-07-20 09:39:14,291  charmcraft.commands.build                DEBUG    Running external command ['env', '-i', 'LANG=C.UTF-8', 'LC_ALL=C.UTF-8', 'PATH=/home/claudio/.venv/charmcraf
t/bin:/home/claudio/.local/bin:/home/claudio/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/bin:/home/claudio/go/bin:/home/cla
udio/.cargo/bin', '/home/claudio/.venv/charmcraft/bin/python', '-I', '/home/claudio/.venv/charmcraft/lib/python3.8/site-packages/charmcraft/charm_builder.py', '--charmdir', '/home/cla
udio/charmcraft/charmcraft/testdir', '--builddir', '/home/claudio/charmcraft/charmcraft/testdir/build', '--entrypoint', '/home/claudio/charmcraft/charmcraft/testdir/src/charm.py', '-r
', '/home/claudio/charmcraft/charmcraft/testdir/requirements.txt']                                                                                                                     
2021-07-20 09:39:14,377  charmcraft.commands.build                DEBUG    :: Starting charm builder                                                                                   
2021-07-20 09:39:14,377  charmcraft.commands.build                DEBUG    :: Translated .jujuignore 2 ".git" => ".*/\.git\Z"                                                          
2021-07-20 09:39:14,378  charmcraft.commands.build                DEBUG    :: Translated .jujuignore 3 ".svn" => ".*/\.svn\Z"                                                          
2021-07-20 09:39:14,378  charmcraft.commands.build                DEBUG    :: Translated .jujuignore 4 ".hg" => ".*/\.hg\Z"                                                            
2021-07-20 09:39:14,378  charmcraft.commands.build                DEBUG    :: Translated .jujuignore 5 ".bzr" => ".*/\.bzr\Z"                                                          
2021-07-20 09:39:14,378  charmcraft.commands.build                DEBUG    :: Translated .jujuignore 6 ".tox" => ".*/\.tox\Z"
```
If we add an extra timestamp, it will be logged as:
```
2021-07-20 09:43:14,055  charmcraft.commands.build                DEBUG    Building for 'bases[0]' as host matches 'build-on[0]'.
2021-07-20 09:43:14,055  charmcraft.commands.build                DEBUG    Building charm in '/home/claudio/charmcraft/charmcraft/testdir/build'
2021-07-20 09:43:14,055  charmcraft.commands.build                DEBUG    Running external command ['env', '-i', 'LANG=C.UTF-8', 'LC_ALL=C.UTF-8', 'PATH=/home/claudio/.venv/charmcraf
t/bin:/home/claudio/.local/bin:/home/claudio/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/bin:/home/claudio/go/bin:/home/cla
udio/.cargo/bin', '/home/claudio/.venv/charmcraft/bin/python', '-I', '/home/claudio/.venv/charmcraft/lib/python3.8/site-packages/charmcraft/charm_builder.py', '--charmdir', '/home/cla
udio/charmcraft/charmcraft/testdir', '--builddir', '/home/claudio/charmcraft/charmcraft/testdir/build', '--entrypoint', '/home/claudio/charmcraft/charmcraft/testdir/src/charm.py', '-r
', '/home/claudio/charmcraft/charmcraft/testdir/requirements.txt']
2021-07-20 09:43:14,144  charmcraft.commands.build                DEBUG    :: 2021-07-20 09:43:14,143  __main__                 Starting charm builder
2021-07-20 09:43:14,144  charmcraft.commands.build                DEBUG    :: 2021-07-20 09:43:14,144  charmcraft.jujuignore    Translated .jujuignore 2 ".git" => ".*/\.git\Z"
2021-07-20 09:43:14,144  charmcraft.commands.build                DEBUG    :: 2021-07-20 09:43:14,144  charmcraft.jujuignore    Translated .jujuignore 3 ".svn" => ".*/\.svn\Z"
2021-07-20 09:43:14,144  charmcraft.commands.build                DEBUG    :: 2021-07-20 09:43:14,144  charmcraft.jujuignore    Translated .jujuignore 4 ".hg" => ".*/\.hg\Z"
2021-07-20 09:43:14,144  charmcraft.commands.build                DEBUG    :: 2021-07-20 09:43:14,144  charmcraft.jujuignore    Translated .jujuignore 5 ".bzr" => ".*/\.bzr\Z"
2021-07-20 09:43:14,144  charmcraft.commands.build                DEBUG    :: 2021-07-20 09:43:14,144  charmcraft.jujuignore    Translated .jujuignore 6 ".tox" => ".*/\.tox\Z"
```
Please confirm that this is what you want, and I'll do it. Maybe just adding `%(name)` instead would be good enough?

Reply:
No, if we're always calling it from a subprocess, we're fine.